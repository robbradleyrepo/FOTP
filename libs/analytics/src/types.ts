export interface AnalyticsConfig {
  integrations?: {
    googleTagManager?: {
      key: string;
      proxyUrl?: string;
    };
    googleOptimize?: {
      key: string;
    };
  };
}
