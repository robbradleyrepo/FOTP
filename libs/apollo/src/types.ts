import type { NormalizedCacheObject } from "apollo-cache-inmemory";
import type { ApolloClient } from "apollo-client";
import type { NextComponentType, NextPageContext } from "next";

export type ApolloClientType = ApolloClient<NormalizedCacheObject>;

// eslint-disable-next-line @typescript-eslint/ban-types
export type NextPageWithApollo<P = {}, IP = P> = NextComponentType<
  NextPageWithApolloContext,
  IP,
  P
>;

export interface NextPageWithApolloContext extends NextPageContext {
  apolloClient: ApolloClientType;
}
