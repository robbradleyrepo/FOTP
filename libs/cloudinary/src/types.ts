export interface CloudinaryConfig {
  defaultCloudName: string;
  prefix?: string;
}

// https://cloudinary.com/documentation/image_transformation_reference#flags_parameter
export enum CloudinaryFlag {
  ANY_FORMAT = "any_format",
  ATTACHMENT = "attachment",
  APNG = "apng",
  AWEBP = "awebp",
  CLIP = "clip",
  CLIP_EVENODD = "clip_evenodd",
  CUTTER = "cutter",
  FORCE_STRIP = "force_strip",
  FORCE_ICC = "force_icc",
  GETINFO = "getinfo",
  IGNORE_ASPECT_RATIO = "ignore_aspect_ratio",
  IGNORE_MASK_CHANNELS = "ignore_mask_channels",
  IMMUTABLE_CACHE = "immutable_cache",
  KEEP_ATTRIBUTION = "keep_attribution",
  KEEP_IPTC = "keep_iptc",
  LAYER_APPLY = "layer_apply",
  LOSSY = "lossy",
  NO_OVERFLOW = "no_overflow",
  PRESERVE_TRANSPARENCY = "preserve_transparency",
  PNG8 = "png8",
  PNG24 = "png24",
  PNG32 = "png32",
  PROGRESSIVE = "progressive",
  PROGRESSIVE_SEMI = "progressive:semi",
  PROGRESSIVE_STEEP = "progressive:steep",
  PROGRESSIVE_NONE = "progressive:none",
  RASTERIZE = "rasterize",
  REGION_RELATIVE = "region_relative",
  RELATIVE = "relative",
  REPLACE_IMAGE = "replace_image",
  SANITIZE = "sanitize",
  STRIP_PROFILE = "strip_profile",
  TEXT_NO_TRIM = "text_no_trim",
  TEXT_DISALLOW_OVERFLOW = "text_disallow_overflow",
  TIFF8_LZW = "tiff8_lzw",
  TILED = "tiled",
}

export enum CloudinaryMedia {
  IMAGE = "image",
  VIDEO = "video",
}

export interface CloudinaryUrlParams {
  cloudName?: string;
  flags?: CloudinaryFlag[];
  format?: string;
  quality?: number | "auto";
  type?: CloudinaryMedia;
  width?: number | "auto";
}

export type CloudinaryFetchedImageSrcSetParams = Omit<
  CloudinaryFetchedImageUrlParams,
  "width"
> &
  Record<"widths", number[]>;

export interface CloudinaryFetchedImageUrlParams extends CloudinaryUrlParams {
  url: string;
}

export type CloudinaryUploadSrcSetParams = Omit<
  CloudinaryUploadUrlParams,
  "width"
> &
  Record<"widths", number[]>;

export interface CloudinaryUploadUrlParams extends CloudinaryUrlParams {
  publicId: string;
  version: string;
}
