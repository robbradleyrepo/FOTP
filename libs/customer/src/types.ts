import { Traits } from "./traits";

export type Customer = Traits;
