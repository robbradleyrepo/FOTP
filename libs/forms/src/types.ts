import { AnyObject } from "final-form";
import { ReactNode } from "react";

export type FieldErrors = ReactNode[];

export type FormErrors = Record<string, FieldErrors>;

export type Validator = (
  value: unknown,
  values: AnyObject
) => FieldErrors | undefined;
