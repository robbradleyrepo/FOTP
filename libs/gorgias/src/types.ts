// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../types/gorgias.d.ts" />
export interface GorgiasConfig {
  appId?: string;
}

export enum GorgiasLocalStorageKey {
  CHAT_ACCESS_TOKEN = "gorgias.chat-access-token",
  CHAT_LAST_PENDING_MESSAGE = "gorgias.chat-last-pending-message",
  CHAT_LAST_SEEN = "gorgias.chat-last-seen",
  EMAIL_CAPTURED = "gorgias.email-captured",
  SEEN_CAMPAIGNS_IDS = "gorgias.seen-campaigns-ids",
  SSP = "gorgias.ssp",
}

export interface GorgiasExternalConfig {
  application: Required<Window["GORGIAS_CHAT_APP"]>;
  bundleVersion: Required<Window["GORGIAS_CHAT_BUNDLE_VERSION"]>;
  texts: Window["GORGIAS_CHAT_TEXTS"];
}
