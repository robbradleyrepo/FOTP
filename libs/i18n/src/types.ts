import { ResourceLanguage } from "i18next";

export interface I18nConfig {
  fallbackLocale: Locale;
  supportedLocales: readonly Locale[];
}

export interface Locale {
  readonly countryCode: string;
  readonly label: string;
  readonly langtag: string;
  readonly language: string;
  readonly region: string;
  readonly resource: ResourceLanguage;
  readonly timeZone: string;
}
