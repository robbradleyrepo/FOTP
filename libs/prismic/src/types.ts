import {
  DocumentHyperlinkData,
  Elements,
  FileHyperlinkData,
  HyperlinkData,
  ImageHyperlinkData,
  RichTextBlock,
  WebHyperlinkData,
} from "prismic-reactjs";
import { ReactElement } from "react";

export { Elements };
export type {
  DocumentHyperlinkData,
  FileHyperlinkData,
  HyperlinkData,
  ImageHyperlinkData,
  RichTextBlock,
  WebHyperlinkData,
};

export enum ArticlePageTag {
  BLOG = "Blog",
  CONDITION = "Condition",
  EVIDENCE = "Evidence",
  INGREDIENT = "Ingredient",
  SUPPLEMENT = "Supplement",
}

export type ContentSlice =
  | ImageSlice
  | QuoteSlice
  | RichTextSlice
  | YouTubeSlice;

export enum ContentSliceType {
  IMAGE = "image",
  QUOTE = "quote",
  RICH_TEXT = "rich_text",
  YOUTUBE = "youtube",
}

export interface Customization {
  body: CustomizationSlice[];
}

export interface CustomizationDict {
  [CustomizationSliceType.IMAGE]: Record<string, ImageCustomizationSlice>;
  [CustomizationSliceType.RICH_TEXT]: Record<
    string,
    RichTextCustomizationSlice
  >;
  [CustomizationSliceType.TITLE]: Record<string, TitleCustomizationSlice>;
}

export type CustomizationSlice =
  | ImageCustomizationSlice
  | RichTextCustomizationSlice
  | TitleCustomizationSlice;

export enum CustomizationSliceType {
  IMAGE = "image_customization",
  RICH_TEXT = "rich_text_customization",
  TITLE = "title_customization",
}

export interface DocumentLink {
  _linkType: LinkType.document;
  _meta: Omit<Meta, "lastPublicationDate">;
}

export type ElementRendererDict = Record<Elements, ReactElement | null>;

export interface FileLink {
  _linkType: LinkType.file;
  name: string;
  size: number;
  url: string;
}

export enum HyperlinkType {
  document = "Document",
  media = "Media",
  web = "Web",
}

export interface Image {
  alt: string | null;
  dimensions: {
    height: number;
    width: number;
  };
  url: string;
}

export interface ImageCustomizationSlice extends Slice {
  primary: {
    image: Image | null;
    target: string | null;
  };
  type: CustomizationSliceType.IMAGE;
}

export interface ImageLink {
  _linkType: LinkType.image;
  name: string;
  url: string;
  size: number;
  height: number;
  width: number;
}

export interface ImageSlice extends Slice {
  primary: {
    caption: RichTextBlock[] | null;
    image: Image | null;
    link: Link | null;
  };
  type: ContentSliceType.IMAGE;
}

export interface LatLng {
  latitude: number;
  longitude: number;
}

export type Link = DocumentLink | FileLink | ImageLink | WebLink;

export enum LinkType {
  document = "Link.document",
  file = "Link.file",
  image = "Link.image",
  web = "Link.web",
}

export interface Meta {
  id: string;
  lang: string;
  lastPublicationDate?: string;
  tags: string[];
  type: string;
  uid?: string;
}

export interface PageInfo {
  endCursor: string;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string;
}

export interface Person extends PrismicDocument {
  image: Image | null;
  name: RichTextBlock[] | null;
  role: RichTextBlock[] | null;
}

export interface PrismicConfig {
  baseUrl: string;
  links: Record<string, string>;
}

export interface PrismicDocument {
  _meta: Meta;
}

export interface QuoteSlice extends Slice {
  label: string | null;
  primary: Record<"attribution" | "quote", RichTextBlock[] | null>;
  type: ContentSliceType.QUOTE;
}

export interface Review extends PrismicDocument {
  body: RichTextBlock[] | null;
  easeOfUse: number | null;
  effectiveness: number | null;
  highlight: RichTextBlock[] | null;
  image: Image | null;
  rating: number | null;
  reviewer: RichTextBlock[] | null;
  title: RichTextBlock[] | null;
}

export interface RichTextSlice extends Slice {
  label: string | null;
  primary: Record<"content", RichTextBlock[]>;
  type: ContentSliceType.RICH_TEXT;
}

export type ResponsiveImage<K extends string> = Image &
  Partial<Record<K, Image>>;

export interface RichTextCustomizationSlice extends Slice {
  primary: {
    richText: RichTextBlock[] | null;
    target: string | null;
  };
  type: CustomizationSliceType.RICH_TEXT;
}

export interface Similar {
  documentId: string;
  max: number;
}
export interface Slice {
  type: string;
}

export type StrictMeta = Meta & Required<Pick<Meta, "uid">>;

export interface TitleCustomizationSlice extends Slice {
  primary: {
    target: string | null;
    title: RichTextBlock[] | null;
  };
  type: CustomizationSliceType.TITLE;
}

export interface YouTubeSlice extends Slice {
  primary: Record<"videoId", string>;
  type: ContentSliceType.YOUTUBE;
}

export interface WebLink {
  _linkType: LinkType.web;
  url: string;
}
