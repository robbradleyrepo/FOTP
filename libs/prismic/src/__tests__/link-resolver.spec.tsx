import { captureException } from "@sentry/nextjs";
import { Link as InternalLink } from "@sss/next";
import React from "react";
import { create } from "react-test-renderer";
import { mocked } from "ts-jest/utils";

import {
  documentResolver,
  linkNormalizer,
  LinkResolver,
  linkResolver,
} from "../link-resolver";
import { ArticlePageTag, HyperlinkData, Link, LinkType } from "../types";

jest.mock("@sentry/nextjs", () => ({
  captureException: jest.fn(),
}));
jest.mock("@sss/i18n", () => ({
  Link: () => null,
}));

const mockedCaptureException = mocked(captureException);

const NODE_ENV = process.env.NODE_ENV;

afterEach(() => {
  jest.clearAllMocks();

  (process.env.NODE_ENV as unknown) = NODE_ENV;
});

describe("documentResolver", () => {
  describe("tag-based routes", () => {
    describe("article_page", () => {
      it("should map all known tag types", () => {
        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [ArticlePageTag.BLOG],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/blog/foo");

        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [ArticlePageTag.CONDITION],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/condition/foo");

        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [ArticlePageTag.INGREDIENT],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/ingredient/foo");

        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [ArticlePageTag.SUPPLEMENT],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/supplement/foo");
      });

      it("should map pages with multiple tags, giving priority to the first known tag", () => {
        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [ArticlePageTag.CONDITION, ArticlePageTag.BLOG],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/condition/foo");

        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [
                ArticlePageTag.SUPPLEMENT,
                ArticlePageTag.BLOG,
                ArticlePageTag.CONDITION,
              ],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/supplement/foo");

        expect(
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [
                "unknown",
                ArticlePageTag.INGREDIENT,
                ArticlePageTag.CONDITION,
              ],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toBe("/ingredient/foo");
      });

      it("should throw an error for pages without tags", () => {
        expect(() =>
          documentResolver({
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [],
              type: "article_page",
              uid: "foo",
            },
          })
        ).toThrowErrorMatchingSnapshot();
      });
    });

    describe("education_video", () => {
      it("should build valid path with tag namespace", () => {
        const docId = "education_video_idx";
        const courseTag = "course:abc";
        const courseId = courseTag.replace("course:", "");

        expect(
          documentResolver({
            _meta: {
              id: "0123456789xyz",
              lang: "us-en",
              lastPublicationDate: "2022-02-03T04:05:06.000Z",
              tags: [courseTag],
              type: "education_video",
              uid: docId,
            },
          })
        ).toBe(`/learn/${courseId}/${docId}`);
      });
    });
  });

  describe("handle-based routes", () => {
    [
      "batch_result",
      "expert",
      "landing_page",
      "product",
      "product_page",
    ].forEach((type) =>
      describe(type, () => {
        it("should interpolate the `{{ handle }}` placeholder", () => {
          expect(
            documentResolver({
              _meta: {
                id: "0123456789abcdef",
                lang: "us-en",
                lastPublicationDate: "2022-02-03T04:05:06.000Z",
                tags: [],
                type,
                uid: "foo",
              },
            })
          ).toMatchSnapshot();
        });

        it("should throw an error for pages without a UID", () => {
          expect(() =>
            documentResolver({
              _meta: {
                id: "0123456789abcdef",
                lang: "us-en",
                lastPublicationDate: "2022-02-03T04:05:06.000Z",
                tags: [],
                type,
              },
            })
          ).toThrowErrorMatchingSnapshot();
        });
      })
    );
  });

  describe("static routes", () => {
    [
      "evidence_page",
      "faq_page",
      "home_page",
      "ingredients_page",
      "experts_page",
      "testing_page",
    ].forEach((type) =>
      describe(type, () => {
        it("should interpolate the `{{ handle }}` placeholder", () => {
          expect(
            documentResolver({
              _meta: {
                id: "0123456789abcdef",
                lang: "us-en",
                lastPublicationDate: "2022-02-03T04:05:06.000Z",
                tags: [],
                type,
                uid: "foo",
              },
            })
          ).toMatchSnapshot();
        });
      })
    );
  });

  it("should by default throw an error for unmapped documents", () => {
    expect(() =>
      documentResolver({
        _meta: {
          id: "0123456789abcdef",
          lang: "us-en",
          lastPublicationDate: "2022-02-03T04:05:06.000Z",
          tags: [],
          type: "unknown",
          uid: "foo-bar",
        },
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it("should return the fallback unmapped documents when specified", () => {
    expect(
      documentResolver(
        {
          _meta: {
            id: "0123456789abcdef",
            lang: "us-en",
            lastPublicationDate: "2022-02-03T04:05:06.000Z",
            tags: [],
            type: "unknown",
            uid: "foo-bar",
          },
        },
        "/some-fallback-page"
      )
    ).toEqual("/some-fallback-page");
  });
});

describe("linkNormalizer", () => {
  it("should handle Prismic documents", () => {
    expect(
      linkNormalizer({
        id: "XlOmChEAACUAkyRa",
        lang: "us-en",
        link_type: "Document", // eslint-disable-line @typescript-eslint/naming-convention
        tags: ["Foo"],
        type: "product",
        uid: "the-one",
      })
    ).toMatchSnapshot();
  });

  it("should handle files", () => {
    expect(
      linkNormalizer({
        kind: "document",
        link_type: "Media", // eslint-disable-line @typescript-eslint/naming-convention
        name: "mv-01620du-genista-potency.pdf",
        size: "204041",
        url:
          "https://prod-fotp-frontend.cdn.prismic.io/prod-fotp-frontend/bc2b9493-eafa-4394-b554-39dfe01f6cd4_mv-01620du-genista-potency.pdf",
      })
    ).toMatchSnapshot();
  });

  it("should handle images", () => {
    expect(
      linkNormalizer({
        height: "600",
        kind: "image",
        link_type: "Media", // eslint-disable-line @typescript-eslint/naming-convention
        name: "harmony-feature-subscribe-and-save.png",
        size: "176120",
        url:
          "https://images.prismic.io/prod-fotp-frontend/c142111c-666f-461a-ba43-f2e3af90edab_harmony-feature-subscribe-and-save.png?auto=compress,format",
        width: "737",
      })
    ).toMatchSnapshot();
  });

  it("should handle external links", () => {
    expect(
      linkNormalizer({
        link_type: "Web", // eslint-disable-line @typescript-eslint/naming-convention
        url: "https://us.fotp.com/",
      })
    ).toMatchSnapshot();
  });

  it("should throw on unknown link types", () => {
    expect(() =>
      linkNormalizer(({
        link_type: "Foo", // eslint-disable-line @typescript-eslint/naming-convention
      } as unknown) as HyperlinkData)
    ).toThrowErrorMatchingSnapshot();
  });
});

describe("linkResolver", () => {
  it("should map Prismic documents to their corresponding pages", () => {
    [
      {
        id: "0123456789abcdef",
        lang: "us-en",
        lastPublicationDate: "2022-02-03T04:05:06.000Z",
        tags: [],
        type: "product",
        uid: "foo-bar",
      },
      {
        id: "0123456789abcdef",
        lang: "us-en",
        lastPublicationDate: "2022-02-03T04:05:06.000Z",
        tags: [],
        type: "ingredients_page",
        uid: "not-used",
      },
    ].forEach((_meta) => {
      expect(
        linkResolver({
          _linkType: LinkType.document,
          _meta,
        })
      ).toBe(documentResolver({ _meta }));
    });
  });

  it("should convert internal `web` links to relative paths", () => {
    expect(
      linkResolver({
        _linkType: LinkType.web,
        url: "https://fotp.com",
      })
    ).toBe("/");

    expect(
      linkResolver({
        _linkType: LinkType.web,
        url: "https://fotp.com/foo/bar/baz",
      })
    ).toBe("/foo/bar/baz");
  });

  it("should remove legacy locale prefixes from internal `web` links", () => {
    expect(
      linkResolver({
        _linkType: LinkType.web,
        url: "https://fotp.com/us/en",
      })
    ).toBe("/");

    expect(
      linkResolver({
        _linkType: LinkType.web,
        url: "https://fotp.com/us/en/foo/bar/baz",
      })
    ).toBe("/foo/bar/baz");
  });

  it("should map media and external `web` links to their URLs", () => {
    expect(
      linkResolver({
        _linkType: LinkType.file,
        name: "Test",
        size: 123,
        url: "https://foo.bar/path/to/file.pdf",
      })
    ).toBe("https://foo.bar/path/to/file.pdf");

    expect(
      linkResolver({
        _linkType: LinkType.image,
        height: 456,
        name: "Test",
        size: 123,
        url: "https://foo.bar/path/to/image.jpg",
        width: 789,
      })
    ).toBe("https://foo.bar/path/to/image.jpg");

    expect(
      linkResolver({
        _linkType: LinkType.web,
        url: "https://foo.bar/",
      })
    ).toBe("https://foo.bar/");
  });
});

describe("<LinkResolver />", () => {
  it("should render mapped documents using an `InternalLink` or the provided `internal` component, with the mapped URL as its `to` prop", () => {
    const link: Link = {
      _linkType: LinkType.document,
      _meta: {
        id: "0123456789abcdef",
        lang: "us-en",
        tags: [],
        type: "product",
        uid: "foo-bar",
      },
    };

    const components = {
      internal: function InternalLink() {
        return null;
      },
    };
    const to = linkResolver(link);

    const defaultLink = create(<LinkResolver link={link} />).root.findByType(
      InternalLink
    );

    expect(defaultLink.props.to).toBe(to);

    const customLink = create(
      <LinkResolver components={components} link={link} />
    ).root.findByType(components.internal);

    expect(customLink.props.to).toBe(to);
  });

  it("should throw an error if a `link` is not provided when not in production", () => {
    expect(() => create(<LinkResolver />)).toThrowError("Missing `link` data");
  });

  it("should render using a `span` or the provided `fallback` component and log the error if a `link` is not provided in production", () => {
    (process.env.NODE_ENV as unknown) = "production";

    const components = {
      fallback: function Fallback() {
        return null;
      },
    };

    expect(create(<LinkResolver />).root.findByType("span")).toBeTruthy();

    expect(mockedCaptureException).toHaveBeenCalledWith(
      expect.objectContaining(new Error("Error: Missing `link` data")),
      { extra: { link: undefined } }
    );

    expect(
      create(<LinkResolver components={components} />).root.findByType(
        components.fallback
      )
    ).toBeTruthy();
  });

  it("should throw an error for unmapped documents when not in production", () => {
    expect(() =>
      create(
        <LinkResolver
          link={{
            _linkType: LinkType.document,
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              tags: [],
              type: "unknown",
              uid: "foo-bar",
            },
          }}
        />
      )
    ).toThrowError(
      'Unable to resolve document with type "unknown" and ID "0123456789abcdef"'
    );
  });

  it("should throw an error for unmapped education course video tags when not in production", () => {
    expect(() =>
      create(
        <LinkResolver
          link={{
            _linkType: LinkType.document,
            _meta: {
              id: "0123456789abcdef",
              lang: "us-en",
              tags: ["invalidCourseTag"],
              type: "education_video",
              uid: "foo-bar",
            },
          }}
        />
      )
    ).toThrowError(
      'Missing valid tags on education course episode page "foo-bar"'
    );
  });

  it("should render unmapped documents using a `span` or the provided `fallback` component and log the error in production", () => {
    (process.env.NODE_ENV as unknown) = "production";

    const link: Link = {
      _linkType: LinkType.document,
      _meta: {
        id: "0123456789abcdef",
        lang: "us-en",
        tags: [],
        type: "unknown",
        uid: "foo-bar",
      },
    };

    const components = {
      fallback: function Fallback() {
        return null;
      },
    };

    expect(
      create(<LinkResolver link={link} />).root.findByType("span")
    ).toBeTruthy();
    expect(
      create(
        <LinkResolver components={components} link={link} />
      ).root.findByType(components.fallback)
    ).toBeTruthy();

    expect(mockedCaptureException).toHaveBeenCalledWith(
      expect.objectContaining(
        new Error(
          'Unable to resolve document with type "unknown" and ID "0123456789abcdef"'
        )
      ),
      { extra: { link } }
    );
  });

  it("should render media and external links using an `a` or the provided `external` component", () => {
    const links: Link[] = [
      {
        _linkType: LinkType.file,
        name: "Test",
        size: 123,
        url: "https://foo.bar/file.pdf",
      },
      {
        _linkType: LinkType.image,
        height: 456,
        name: "Test",
        size: 123,
        url: "https://foo.bar/image.jpg",
        width: 789,
      },
      {
        _linkType: LinkType.web,
        url: "https://foo.bar/",
      },
    ];

    links.forEach((link) => {
      const components = {
        external: function ExternalLink() {
          return null;
        },
      };
      const href = linkResolver(link);

      const defaultLink = create(<LinkResolver link={link} />).root.findByType(
        "a"
      );

      expect(defaultLink.props.href).toBe(href);

      const customLink = create(
        <LinkResolver components={components} link={link} />
      ).root.findByType(components.external);

      expect(customLink.props.href).toBe(href);
    });
  });
});
