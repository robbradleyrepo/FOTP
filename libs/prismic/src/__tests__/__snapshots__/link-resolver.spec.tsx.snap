// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentResolver handle-based routes batch_result should interpolate the \`{{ handle }}\` placeholder 1`] = `"/science/testing-and-transparency/foo"`;

exports[`documentResolver handle-based routes batch_result should throw an error for pages without a UID 1`] = `"Missing UID for document type \\"batch_result\\" on document ID \\"0123456789abcdef\\""`;

exports[`documentResolver handle-based routes expert should interpolate the \`{{ handle }}\` placeholder 1`] = `"/science/experts/foo"`;

exports[`documentResolver handle-based routes expert should throw an error for pages without a UID 1`] = `"Missing UID for document type \\"expert\\" on document ID \\"0123456789abcdef\\""`;

exports[`documentResolver handle-based routes landing_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/landing/foo"`;

exports[`documentResolver handle-based routes landing_page should throw an error for pages without a UID 1`] = `"Missing UID for document type \\"landing_page\\" on document ID \\"0123456789abcdef\\""`;

exports[`documentResolver handle-based routes product should interpolate the \`{{ handle }}\` placeholder 1`] = `"/products/foo"`;

exports[`documentResolver handle-based routes product should throw an error for pages without a UID 1`] = `"Missing UID for document type \\"product\\" on document ID \\"0123456789abcdef\\""`;

exports[`documentResolver handle-based routes product_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/products/foo"`;

exports[`documentResolver handle-based routes product_page should throw an error for pages without a UID 1`] = `"Missing UID for document type \\"product_page\\" on document ID \\"0123456789abcdef\\""`;

exports[`documentResolver should by default throw an error for unmapped documents 1`] = `"Unable to resolve document with type \\"unknown\\" and ID \\"0123456789abcdef\\""`;

exports[`documentResolver static routes evidence_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"science/evidence"`;

exports[`documentResolver static routes experts_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/science/experts"`;

exports[`documentResolver static routes faq_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/help/faq"`;

exports[`documentResolver static routes home_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/"`;

exports[`documentResolver static routes ingredients_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/science/ingredients"`;

exports[`documentResolver static routes testing_page should interpolate the \`{{ handle }}\` placeholder 1`] = `"/science/testing-and-transparency"`;

exports[`documentResolver tag-based routes article_page should throw an error for pages without tags 1`] = `"Missing tags on article page \\"foo\\""`;

exports[`linkNormalizer should handle Prismic documents 1`] = `
Object {
  "_linkType": "Link.document",
  "_meta": Object {
    "id": "XlOmChEAACUAkyRa",
    "lang": "us-en",
    "tags": Array [
      "Foo",
    ],
    "type": "product",
    "uid": "the-one",
  },
}
`;

exports[`linkNormalizer should handle external links 1`] = `
Object {
  "_linkType": "Link.web",
  "url": "https://us.fotp.com/",
}
`;

exports[`linkNormalizer should handle files 1`] = `
Object {
  "_linkType": "Link.file",
  "name": "mv-01620du-genista-potency.pdf",
  "size": 204041,
  "url": "https://prod-fotp-frontend.cdn.prismic.io/prod-fotp-frontend/bc2b9493-eafa-4394-b554-39dfe01f6cd4_mv-01620du-genista-potency.pdf",
}
`;

exports[`linkNormalizer should handle images 1`] = `
Object {
  "_linkType": "Link.image",
  "height": 600,
  "name": "harmony-feature-subscribe-and-save.png",
  "size": 176120,
  "url": "https://images.prismic.io/prod-fotp-frontend/c142111c-666f-461a-ba43-f2e3af90edab_harmony-feature-subscribe-and-save.png?auto=compress,format",
  "width": 737,
}
`;

exports[`linkNormalizer should throw on unknown link types 1`] = `"Invalid link type supplied: expected \\"Document\\" | \\"Media\\" | \\"Web\\", got \\"Foo\\""`;
